### PR TITLE
Simplified porting from client4 to nclient4

### PR DESCRIPTION
**Pull request by xaionaro**
Related: https://github.com/insomniacslk/dhcp/issues/334

The PR is supposed to reduce costs of migrating from client4 to nclient4.

Changes in [nclient4](https://github.com/insomniacslk/dhcp/blob/master/dhcpv4/nclient4/client.go):

* Published `Logger`. So now somebody can use option `WithLogger` and define own logger or/and to use a builtin logger. And added comment `print _all_ DHCP messages` to guarantee (to an external user) and to remind (to an internal developer) that PrintMessage always prints all messages. So in task https://github.com/insomniacslk/dhcp/issues/334 it's possible to use this loggers instead of changing `Request`, now.
* Fixed the problem with unicast frames. Previously `nclient4` was always sending broadcast frames. But our utility (related to https://github.com/insomniacslk/dhcp/issues/334) requires also support of unicast frames (which are supported in the old package [client4](https://github.com/insomniacslk/dhcp/blob/master/dhcpv4/client4/client.go)).
* Removed function `isErrClosing(...)`. It's not universal through different implementations of [net.PacketConn](https://golang.org/pkg/net/#PacketConn).
* ~~`DiscoverOffer` and `Request` now returns all messages (not only replies, but also requests). So `Request` is now more logically compatible with the old method [`Exchange`](https://github.com/insomniacslk/dhcp/blob/793c0f67c0a9a2b178bfe98223fd2bfc92a115f3/dhcpv4/client4/client.go#L183).~~ (it's requested to move this change to a separate PR)
* Moved errors to `var`-s and `type`-s. Used `%w` instead of `%v` while wrapping an error with `fmt.Errorf`. Wrapped more received errors with `fmt.Errorf`.

### CODE REVIEW

File: dhcpv4/client4/client.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// This package is deprecated. Use "nclient4" instead.
+
```
Comment (hugelgupf) [inline dhcpv4/client4/client.go]: Remove empty line, so this becomes a godoc.

File: dhcpv4/nclient4/client.go [hunk 2]
```diff
@@ -249,54 +303,57 @@ func (c *Client) receiveLoop() {
...
 func WithTimeout(d time.Duration) ClientOpt {
-	return func(c *Client) {
+	return func(c *Client) (err error) {
 		c.timeout = d
+		return
 	}
 }
 
-// WithSummaryLogger logs one-line DHCPv4 message summarys when sent & received.
-func WithSummaryLogger() ClientOpt {
```
Comment (hugelgupf) [inline dhcpv4/nclient4/client.go]: Please keep the WithSummaryLogger and WithDebugLogger for people already using nclient4.

File: dhcpv4/nclient4/client.go [hunk 3]
```diff
@@ -193,21 +249,19 @@ func (c *Client) Close() error {
 	return err
 }
 
-func isErrClosing(err error) bool {
-	// Unfortunately, the epoll-connection-closed error is internal to the
-	// net library.
-	return strings.Contains(err.Error(), "use of closed network connection")
+func (c *Client) isClosed() bool {
+	return atomic.LoadUint32(&c.closed) != 0
```
Comment (hugelgupf) [inline dhcpv4/nclient4/client.go]: closed is never set? Also, why the change?

File: dhcpv4/nclient4/client.go [hunk 4]
```diff
@@ -249,54 +303,57 @@ func (c *Client) receiveLoop() {
...
 
-// WithDebugLogger logs multi-line full DHCPv4 messages when sent & received.
-func WithDebugLogger() ClientOpt {
-	return func(c *Client) {
-		c.logger = debugLogger{
-			Logger: log.New(os.Stderr, "[dhcpv4] ", log.LstdFlags),
-		}
+// WithUnicast forces client to send messages as unicast frames.
+// By default client sends messages as broadcast frames even if server address is defined.
+func WithUnicast() ClientOpt {
```
Comment (hugelgupf) [inline dhcpv4/nclient4/client.go]: Can you allow users to specify a *net.UDPAddr here?

If they don't specify one, choose the default like you did here?

File: dhcpv4/nclient4/client.go [hunk 5]
```diff
@@ -441,7 +515,8 @@ func (c *Client) retryFn(fn func(timeout time.Duration) error) error {
 	timeout := c.timeout
 
 	// Each retry takes the amount of timeout at worst.
-	for i := 0; i < c.retry || c.retry < 0; i++ {
+	for i := 0; i < c.retry || c.retry < 0; i++ { // TODO: why is this called "retry" if this is "tries" ("retries"+1)?
+		c.logger.Printf("read-timeout is %v", timeout)
```
Comment (hugelgupf) [inline dhcpv4/nclient4/client.go]: revert this debug print?

File: dhcpv4/nclient4/client.go (hunk 3 above)
Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: `c.closed` is set in `c.Close()`. The change because the previous way is:

* Hacky. Search a substring in the error description is an unstrict/risky way to handle errors.
* Wrong. I can implement `net.PacketConn` which returns other error. And this package accepts arbitrary `net.PacketConn`-s via `NewWithConn(...)`.

**Comment by xaionaro** [comment]
Thank you for your time and review :)

I fixed almost everything. Only one thing left for discussion (`isClosed`).

**Comment by pmazzini** [comment]
Please sign the commit:

```
git rebase HEAD~2 --signoff
git push --force-with-lease origin master
```

> DiscoverOffer and Request now returns all messages (not only replies, but also requests). So Request is now more logically compatible with the old method Exchange.

Can this be done on a separate PR?

> Used xerrors.Errorf instead of fmt.Errorf

This will introduce a new dependency to `u-root`. Can't we use something from the standard library instead?

**Comment by xaionaro** [comment]
> Can't we use something from the standard library instead?

Yeah. It appears they already added flag `%w` to `fmt.Errorf`, so I just could change `%v` to `%w`. Will be fixed.

> If the format specifier includes a %w verb with an error operand, the returned error will implement an Unwrap method returning the operand. It is invalid to include more than one %w verb or to supply it with an operand that does not implement the error interface. The %w verb is otherwise a synonym for %v.

**Comment by xaionaro** [comment]
@pmazzini : removed an extra dependency, squashed commits, signed-off :)

**Comment by xaionaro** [comment]
@hugelgupf : I used `%w` for `fmt.Errorf`. So it will print neat error description only on Go>1.13.

For comparison:

An error description on Go1.13
```
could not get local IPv4 from fbwhoami: host doesn't have an IPv4 address
```

An error description on Go1.12:
```
could not get local IPv4 from fbwhoami: &{%!w(string=host doesn't have an IPv4 address)}
```

I hope it's not a problem.

File: .travis/tests.sh [hunk 6]
```diff
@@ -10,14 +10,20 @@ echo "" > coverage.txt
 # tests.
 ip a
 
+GO_TEST_OPTS=()
+if [[ "$TRAVIS_GO_VERSION" =~ ^1.(9|10|11|12)$ ]]
+then
+    GO_TEST_OPTS+='-vet=off'
```
Comment (pmazzini) [inline .travis/tests.sh]: Maybe add a comment that this is because we are using the `%w` verb.

File: dhcpv4/nclient4/client.go [hunk 7]
```diff
@@ -63,35 +74,61 @@ type pendingCh struct {
...
+	// PrintMessage print _all_ DHCP messages
 	PrintMessage(prefix string, message *dhcpv4.DHCPv4)
+
+	// Printf is use to print the rest debugging information
+	Printf(format string, v ...interface{})
 }
 
-type emptyLogger struct{}
+// EmptyLogger prints nothing
+type EmptyLogger struct{}
```
Comment (pmazzini) [inline dhcpv4/nclient4/client.go]: Do we need to make this public?

File: dhcpv4/nclient4/client.go [hunk 8]
```diff
@@ -193,21 +249,19 @@ func (c *Client) Close() error {
...
+	return atomic.LoadUint32(&c.closed) != 0
 }
 
 func (c *Client) receiveLoop() {
 	defer c.wg.Done()
 	for {
 		// TODO: Clients can send a "max packet size" option in their
-		// packets, IIRC. Choose a reasonable size and set it.
-		b := make([]byte, maxMessageSize)
+		//       packets, IIRC. Choose a reasonable size and set it.
```
Comment (pmazzini) [inline dhcpv4/nclient4/client.go]: Is this indentation on purpose?

File: dhcpv4/nclient4/client.go (hunk 7 above)
Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: I don't see any reason to hide it. Somebody can do now something like:
```
var logger nclient4.Logger
switch loggerType {
case someValue0:
    logger = &nclient4.EmptyLogger{}
case someValue1:
    logger = &nclient4.DebugLogger{Printfer: log.New(...)}
case someValue2:
    logger = &myLogger{}
}
opts = append(opts, nclient4.WithLogger(logger))
```

But I can hide it if you want :)

File: dhcpv4/nclient4/client.go [hunk 9]
```diff
@@ -129,39 +166,58 @@ type Client struct {
...
-	i, err := net.InterfaceByName(iface)
-	if err != nil {
-		return nil, err
-	}
-	pc, err := NewRawUDPConn(iface, ClientPort)
-	if err != nil {
-		return nil, err
-	}
-	return NewWithConn(pc, i.HardwareAddr, opts...)
+	return new(iface, nil, nil, opts...)
```
Comment (pmazzini) [inline dhcpv4/nclient4/client.go]: Just out of curiosity, is there a reason for reorganizing this?

File: dhcpv4/nclient4/client.go (hunk 7 above)
Comment (pmazzini) [reply dhcpv4/nclient4/client.go]: No, it is fine, if you think it may be useful lets leave it.

File: dhcpv4/nclient4/client.go [hunk 10]
```diff
@@ -63,35 +74,61 @@ type pendingCh struct {
...
 
-func (e emptyLogger) Printf(format string, v ...interface{})             {}
-func (e emptyLogger) PrintMessage(prefix string, message *dhcpv4.DHCPv4) {}
+// PrintMessage is just a dummy function that does nothing
+func (e EmptyLogger) PrintMessage(prefix string, message *dhcpv4.DHCPv4) {}
 
-type shortSummaryLogger struct {
-	*log.Logger
+// Printfer is used for actual output of the logger. For example *log.Logger is a Printfer.
+type Printfer interface {
```
Comment (pmazzini) [inline dhcpv4/nclient4/client.go]: Do we actually need this new interface?

File: dhcpv4/nclient4/client.go (hunk 9 above)
Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: Yes. The `Conn` should be initialized after processing options to be able to initialize an unicast `Conn` via these options.

File: dhcpv4/nclient4/client.go (hunk 10 above)
Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: I use the same interface in multiple `struct`-s, so it's just DRY. And it's public, because `DebugLogger` and `ShortSummaryLogger` are public.

File: dhcpv4/nclient4/client.go (hunk 8 above)
Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: Yeah (my IDE did not recognize the second line as a "TODO"). But I'll revert it.

File: dhcpv4/nclient4/client.go (hunk 9 above)
Comment (pmazzini) [reply dhcpv4/nclient4/client.go]: Thanks for the clarification!

File: .travis/tests.sh (hunk 6 above)
Comment (xaionaro) [reply .travis/tests.sh]: Fixed.

File: dhcpv4/nclient4/client.go (hunk 8 above)
Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: Fixed.

File: .travis/tests.sh (hunk 6 above)
Comment (xaionaro) [reply .travis/tests.sh]: Fixed.

File: dhcpv4/nclient4/client.go (hunk 10 above)
Comment (pmazzini) [reply dhcpv4/nclient4/client.go]: Why not use `*log.Logger` inside `DebugLogger` and `ShortSummaryLogger`?

Comment (xaionaro) [reply dhcpv4/nclient4/client.go]: "Require less, provide more"

We don't need exactly `*log.Logger`. We need just some `Printf`-er here (not more). Somebody else could use `logrus` or whatever else instead of `*log.Logger`. If you require only the thing you really need you make this library much more flexible. There's only one real disadvantage I know: handling through interfaces is slower than handling directly through an implementation.

**Comment by pmazzini** [comment]
Lets move forward, we can revisit how the close check is being done.
